File: backend/src/services/bedrock/bedrock.service.ts
import { BedrockRuntimeClient } from "@aws-sdk/client-bedrock-runtime";
import { BedrockClient } from "@aws-sdk/client-bedrock";
import { InvokeModelCommand, InvokeModelWithResponseStreamCommand } from "@aws-sdk/client-bedrock-runtime";
import { ListFoundationModelsCommand, ModelModality } from "@aws-sdk/client-bedrock";
import { CostExplorerClient, GetCostAndUsageCommand } from "@aws-sdk/client-cost-explorer";

import {
  AIModelInfo,
  ModelMessage,
  ModelResponse,
  StreamCallbacks,
  ProviderInfo,
  UsageCostInfo,
  ServiceCostInfo,
  InvokeModelParamsRequest,
} from "../../types/ai.types";
import { ApiProvider } from "../../types/ai.types";
import BedrockModelConfigs from "../../config/data/bedrock-models-config.json";
import { createLogger } from "@/utils/logger";
import { getErrorMessage } from "@/utils/errors";
import { BaseProviderService } from "../base.provider";
import { ConnectionParams } from "@/middleware/auth.middleware";

const logger = createLogger(__filename);

interface BedrockModelConfigRecord {
  provider: string;
  modelId: string;
  modelIdOverride?: Record<string, string>;
  name: string;
  regions: string[];
}

export class BedrockService extends BaseProviderService {
  protected bedrockClient: BedrockRuntimeClient;
  protected bedrockManagementClient: BedrockClient;

  constructor(connection: ConnectionParams) {
    super(connection);

    if (!connection.AWS_REGION) {
      logger.warn("AWS_REGION is not set. Skipping AWS Bedrock initialization.");
      return;
    }

    const config = {
      region: connection.AWS_REGION,
      profile: connection.AWS_PROFILE || "", // Use AWS profile if set
      credentials: connection.AWS_PROFILE
        ? undefined
        : {
            accessKeyId: connection.AWS_ACCESS_KEY_ID || "",
            secretAccessKey: connection.AWS_SECRET_ACCESS_KEY || "",
          },
    };

    // AWS Bedrock client configuration
    this.bedrockClient = new BedrockRuntimeClient(config);

    // AWS Bedrock management client for non-runtime operations (listing models, etc.)
    this.bedrockManagementClient = new BedrockClient(config);
  }

  async invokeModel(request: InvokeModelParamsRequest): Promise<ModelResponse> {
    if (!this.bedrockClient) {
      throw new Error("AWS Bedrock client is not initialized. Please check your AWS credentials and region.");
    }

    // Get provider service and parameters
    const { service, params } = await this.formatProviderParams(request);

    // Send command using Bedrock client
    const command = new InvokeModelCommand(params);
    const response = await this.bedrockClient.send(command);

    // Parse the response body
    const responseBody = JSON.parse(new TextDecoder().decode(response.body));
    return service.parseResponse(responseBody, request);
  }

  // Stream response from models using InvokeModelWithResponseStreamCommand
  async invokeModelAsync(request: InvokeModelParamsRequest, callbacks: StreamCallbacks): Promise<void> {
    const { modelId } = request;
    callbacks.onStart?.();

    const provider = this.getModelProvider(modelId);

    // Check if modelId supports streaming (Anthropic, Amazon, Mistral)
    const supportsStreaming = provider === "anthropic" || provider === "amazon" || provider === "mistral";

    if (!supportsStreaming) {
      try {
        // For models that don't support streaming, use the regular generation and simulate streaming
        const response = await this.invokeModel(request);

        // Simulate streaming by sending chunks of the response
        const chunks = response.content.split(" ");
        for (const chunk of chunks) {
          callbacks.onToken?.(chunk + " ");
          // Add a small delay to simulate streaming
          await new Promise(resolve => setTimeout(resolve, 10));
        }

        callbacks.onComplete?.(response.content);
      } catch (e: unknown) {
        logger.error(e, "InvokeModel failed");
        callbacks.onError?.(e instanceof Error ? e : new Error(getErrorMessage(e)));
      }

      return;
    }

    try {
      // Get provider service and parameters
      const { params } = await this.formatProviderParams(request);

      // Create a streaming command
      const streamCommand = new InvokeModelWithResponseStreamCommand(params);
      const streamResponse = await this.bedrockClient.send(streamCommand);

      let fullResponse = "";

      // Process the stream
      if (streamResponse.body) {
        for await (const chunk of streamResponse.body) {
          if (chunk.chunk?.bytes) {
            const decodedChunk = new TextDecoder().decode(chunk.chunk.bytes);
            const chunkData = JSON.parse(decodedChunk);

            // Extract the token based on model provider
            let token = "";
            if (provider === "anthropic") {
              // For Anthropic models
              if (chunkData.type === "content_block_delta" && chunkData.delta?.text) {
                token = chunkData.delta.text;
              }
            } else if (provider === "amazon") {
              // For Amazon models
              if (chunkData.outputText) {
                token = chunkData.outputText;
              }
            } else if (provider === "mistral") {
              // For Mistral models
              if (chunkData.outputs && chunkData.outputs[0]?.text) {
                token = chunkData.outputs[0].text;
              }
            }

            if (token) {
              fullResponse += token;
              callbacks.onToken?.(token);
            }
          }
        }
      }

      callbacks.onComplete?.(fullResponse);
    } catch (e: unknown) {
      logger.error(e, "InvokeModelWithResponseStreamCommand failed");
      callbacks.onError?.(e instanceof Error ? e : new Error(getErrorMessage(e)));
    }
  }

  // Get the appropriate service and parameters based on the model ID
  private async formatProviderParams(request: InvokeModelParamsRequest) {
    const { modelId } = request;

    let service;
    let params;

    let provider = this.getModelProvider(modelId);

    if (provider == "anthropic") {
      const { AnthropicService } = await import("./providers/anthropic.service");
      const anthropicService = new AnthropicService();
      const result = await anthropicService.getInvokeModelParams(request);
      service = anthropicService;
      params = result.params;
    } else if (provider == "amazon") {
      const { AmazonService } = await import("./providers/amazon.service");
      const amazonService = new AmazonService();
      const result = await amazonService.getInvokeModelParams(request);
      service = amazonService;
      params = result.params;
    } else if (provider == "ai21") {
      const { AI21Service } = await import("./providers/ai21.service");
      const ai21Service = new AI21Service();
      const result = await ai21Service.getInvokeModelParams(request);
      service = ai21Service;
      params = result.params;
    } else if (provider == "cohere") {
      const { CohereService } = await import("./providers/cohere.service");
      const cohereService = new CohereService();
      const result = await cohereService.getInvokeModelParams(request);
      service = cohereService;
      params = result.params;
    } else if (provider == "meta") {
      const { MetaService } = await import("./providers/meta.service");
      const metaService = new MetaService();
      const result = await metaService.getInvokeModelParams(request);
      service = metaService;
      params = result.params;
    } else if (provider == "mistral") {
      const { MistralService } = await import("./providers/mistral.service");
      const mistralService = new MistralService();
      const result = await mistralService.getInvokeModelParams(request);
      service = mistralService;
      params = result.params;
    } else {
      throw new Error(`Unsupported model provider: ${provider}`);
    }

    return { service, params };
  }

  getModelProvider(modelId: string) {
    if (modelId.startsWith("us.") || modelId.startsWith("eu.") || modelId.startsWith("ap.")) {
      return modelId.substring(3).split(".")[0];
    }

    return modelId.split(".")[0];
  }

  async getInfo(checkConnection = false): Promise<ProviderInfo> {
    const isConnected = !!this.bedrockClient;
    const region = this.connection.AWS_REGION;
    const profile = this.connection.AWS_PROFILE;
    const accessKey = this.connection.AWS_ACCESS_KEY_ID;

    const details: Record<string, string | number | boolean | undefined> = {
      configured: isConnected,
      credentialsValid: "N/A",
    };

    if (region) details.region = region;
    if (profile) details.profile = profile;
    if (accessKey) details.accessKey = `${accessKey.substring(0, 6)}...`;

    if (isConnected && checkConnection && this.bedrockManagementClient) {
      try {
        // Test credentials by attempting to make a simple API call
        const creds = await this.bedrockManagementClient.config.credentials();
        details.credentialsValid = true;
        if (creds.accountId) details.accountId = creds.accountId;
        if (creds.expiration) details.expiration = creds.expiration?.toISOString();
      } catch (error) {
        logger.error(error, "Error validating AWS credentials");
        details.credentialsValid = false;
        details.errorMessage = getErrorMessage(error);
      }
    }

    return {
      id: ApiProvider.AWS_BEDROCK,
      name: BaseProviderService.getApiProviderName(ApiProvider.AWS_BEDROCK),
      costsInfoAvailable: isConnected,
      isConnected,
      details,
    };
  }

  // Helper method to get all supported models with their metadata
  async getModels(): Promise<Record<string, AIModelInfo>> {
    // no AWS connection
    if (!this.connection.AWS_ACCESS_KEY_ID && !this.connection.AWS_PROFILE) {
      logger.warn("AWS credentials are not set. Skipping AWS Bedrock model retrieval.");
      return {};
    }

    const modelsRegions = (BedrockModelConfigs as BedrockModelConfigRecord[]).reduce(
      (acc: Record<string, string[]>, region) => {
        const { modelId, regions } = region;
        acc[modelId] = regions;
        return acc;
      },
      {}
    );

    const modelIdOverrides = (BedrockModelConfigs as BedrockModelConfigRecord[]).reduce(
      (acc: Record<string, Record<string, string>>, region) => {
        const { modelId, modelIdOverride } = region;
        if (modelIdOverride) {
          acc[modelId] = modelIdOverride;
        }
        return acc;
      },
      {}
    );

    const command = new ListFoundationModelsCommand({});
    const response = await this.bedrockManagementClient.send(command);

    const models: Record<string, any> = {};

    if (!response.modelSummaries || !response.modelSummaries.length) {
      return models;
    }

    const bedrockRegion = await this.bedrockClient.config.region();
    for (const model of response.modelSummaries) {
      const regions = modelsRegions[model.modelId || ""];
      if (!regions || !regions.includes(bedrockRegion)) {
        continue;
      }

      if (model.modelId && model.providerName) {
        let modelId = model.modelId;
        const providerName = model.providerName;

        if (modelIdOverrides[model.modelId]) {
          const map = modelIdOverrides[model.modelId];
          // Use the override if available
          const region = bedrockRegion.split("-")[0]; // Get the region prefix (e.g., "us")
          modelId = map[region] || map[modelId] || modelId;
        }

        models[modelId] = {
          apiProvider: ApiProvider.AWS_BEDROCK,
          provider: providerName,
          name: model.modelName || modelId.split(".").pop() || modelId,
          description: `${model.modelName || modelId} by ${providerName}`,
          supportsStreaming: model.responseStreamingSupported || false,
          supportsTextIn: model.inputModalities?.includes(ModelModality.TEXT) ?? true,
          supportsTextOut: model.outputModalities?.includes(ModelModality.TEXT) ?? true,
          supportsImageIn: model.inputModalities?.includes(ModelModality.IMAGE) || false,
          supportsImageOut: model.outputModalities?.includes(ModelModality.IMAGE) || false,
          supportsEmbeddingsOut: model.outputModalities?.includes(ModelModality.EMBEDDING) || false,
          supportsEmbeddingsIn: model.outputModalities?.includes(ModelModality.EMBEDDING) || false,
        };
      }
    }

    return models;
  }

  async getCosts(startTime: number, endTime?: number): Promise<UsageCostInfo> {
    const result: UsageCostInfo = {
      start: new Date(startTime * 1000),
      end: endTime ? new Date(endTime * 1000) : undefined,
      costs: [],
    };

    if (!this.connection.AWS_ACCESS_KEY_ID && !this.connection.AWS_PROFILE) {
      result.error =
        "AWS credentials are not set. Set AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY or AWS_PROFILE in config.";
      return result;
    }

    try {
      // Create Cost Explorer client using the same region/credentials as Bedrock
      const costExplorerClient = new CostExplorerClient({
        region: this.connection.AWS_REGION,
        credentials: await this.bedrockClient.config.credentials(),
      });

      // Format start and end dates in YYYY-MM-DD format required by Cost Explorer
      const startDate = new Date(startTime * 1000);
      const endDate = endTime ? new Date(endTime * 1000) : new Date();

      const formattedStartDate = startDate.toISOString().split("T")[0];
      const formattedEndDate = endDate.toISOString().split("T")[0];

      // Create command to get cost and usage data
      const command = new GetCostAndUsageCommand({
        TimePeriod: {
          Start: formattedStartDate,
          End: formattedEndDate,
        },
        Granularity: "DAILY",
        Metrics: ["BlendedCost", "UsageQuantity"],
        GroupBy: [
          {
            Type: "DIMENSION",
            Key: "SERVICE",
          },
        ],
      });

      logger.debug({ command }, "Fetching AWS Bedrock costs");

      // Get cost and usage data from AWS Cost Explorer
      const costData = await costExplorerClient.send(command);

      // Process the results into our required format
      const serviceCosts: ServiceCostInfo[] = [];

      // Overall AWS Bedrock cost
      let totalCost = 0;

      // Parse the results
      if (costData.ResultsByTime && costData.ResultsByTime.length > 0) {
        // If we have results, create service cost entries for each group
        for (const result of costData.ResultsByTime) {
          if (result.Groups && result.Groups.length > 0) {
            for (const group of result.Groups) {
              if (group.Metrics && group.Metrics.BlendedCost) {
                const cost = parseFloat(group.Metrics.BlendedCost.Amount || "0");
                totalCost += cost;

                // Extract service name
                const serviceName = group.Keys && group.Keys.length > 0 ? group.Keys[0] : "Amazon Bedrock";

                // Find or create the service cost entry
                let serviceCost = serviceCosts.find(sc => sc.name === serviceName);
                if (!serviceCost) {
                  serviceCost = {
                    name: serviceName,
                    type: "service",
                    amounts: [],
                  };
                  serviceCosts.push(serviceCost);
                }

                // Update or add the amount
                const currency = group.Metrics.BlendedCost.Unit || "USD";
                const existingAmount = serviceCost.amounts.find(a => a.currency === currency);
                if (existingAmount) {
                  existingAmount.amount += cost;
                } else {
                  serviceCost.amounts.push({ amount: cost, currency });
                }
              }
            }
          }
        }
      }

      // Get model specific costs using a more detailed query
      // This is a separate command to get costs grouped by model
      const modelCommand = new GetCostAndUsageCommand({
        TimePeriod: {
          Start: formattedStartDate,
          End: formattedEndDate,
        },
        Granularity: "DAILY",
        Metrics: ["BlendedCost", "UsageQuantity"],
        GroupBy: [
          {
            Type: "DIMENSION",
            Key: "USAGE_TYPE",
          },
        ],
        Filter: {
          Dimensions: {
            Key: "SERVICE",
            Values: ["Amazon Bedrock"],
          },
        },
      });

      const modelCostData = await costExplorerClient.send(modelCommand);

      // Process model-specific costs
      if (modelCostData.ResultsByTime && modelCostData.ResultsByTime.length > 0) {
        for (const result of modelCostData.ResultsByTime) {
          if (result.Groups && result.Groups.length > 0) {
            for (const group of result.Groups) {
              if (group.Metrics && group.Metrics.BlendedCost && group.Keys && group.Keys.length > 0) {
                // Parse usage type to determine model family
                const usageType = group.Keys[0];
                const cost = parseFloat(group.Metrics.BlendedCost.Amount || "0");
                const currency = group.Metrics.BlendedCost.Unit || "USD";

                // Extract model family from usage type
                let modelFamily = "Other Models";

                // Map usage types to model families
                if (usageType.includes("Claude")) {
                  modelFamily = "Anthropic Claude Models";
                } else if (usageType.includes("Titan")) {
                  modelFamily = "Amazon Titan Models";
                } else if (usageType.includes("Jurassic")) {
                  modelFamily = "AI21 Jurassic Models";
                } else if (usageType.includes("Command")) {
                  modelFamily = "Cohere Command Models";
                } else if (usageType.includes("Llama")) {
                  modelFamily = "Meta Llama Models";
                } else if (usageType.includes("Mistral")) {
                  modelFamily = "Mistral AI Models";
                }

                // Find or create the model family cost entry
                let modelCost = serviceCosts.find(sc => sc.name === modelFamily && sc.type === "model_family");
                if (!modelCost) {
                  modelCost = {
                    name: modelFamily,
                    type: "model_family",
                    amounts: [],
                  };
                  serviceCosts.push(modelCost);
                }

                // Update or add the amount
                const existingAmount = modelCost.amounts.find(a => a.currency === currency);
                if (existingAmount) {
                  existingAmount.amount += cost;
                } else {
                  modelCost.amounts.push({ amount: cost, currency });
                }
              }
            }
          }
        }
      }

      // If no detailed model costs were found, but we have total costs,
      // add a generic Amazon Bedrock service entry
      if (serviceCosts.length === 0 && totalCost > 0) {
        serviceCosts.push({
          name: "Amazon Bedrock",
          type: "service",
          amounts: [{ amount: totalCost, currency: "USD" }],
        });
      }

      result.costs = serviceCosts.sort((a, b) => a?.name?.localeCompare(b.name || "") || 0);
      return result;
    } catch (error) {
      logger.error(error, "Error fetching AWS Bedrock usage information");
      result.error = getErrorMessage(error);
      return result;
    }
  }
}

File: backend-rust/src/services/bedrock/bedrock_service.rs
use async_trait::async_trait;
use aws_config::{BehaviorVersion, Region};
use aws_sdk_bedrock::{types::ModelModality, Client as BedrockClient};
use aws_sdk_bedrockruntime::{primitives::Blob, Client as BedrockRuntimeClient};
use chrono::DateTime;
use log::error;
use serde_json::Value;
use std::collections::HashMap;
use std::error::Error;
use std::future::Future;
use std::pin::Pin;

use crate::config::AppConfig;
use crate::services::ai::*;
use crate::services::bedrock::providers::{
    ai21::AI21Provider, amazon::AmazonProvider, anthropic::AnthropicProvider,
    cohere::CohereProvider, meta::MetaProvider, mistral::MistralProvider,
};
use crate::utils::errors::AppError;

#[allow(dead_code)]
pub struct BedrockService {
    config: AppConfig,
    runtime_client: Option<BedrockRuntimeClient>,
    bedrock_client: Option<BedrockClient>,
}

impl BedrockService {
    pub fn new(config: AppConfig) -> Self {
        Self {
            config,
            runtime_client: None,
            bedrock_client: None,
        }
    }

    async fn get_runtime_client(&mut self) -> Result<&BedrockRuntimeClient, AppError> {
        if self.runtime_client.is_none() {
            let aws_config = self.build_aws_config().await?;
            self.runtime_client = Some(BedrockRuntimeClient::new(&aws_config));
        }
        Ok(self.runtime_client.as_ref().unwrap())
    }

    async fn get_bedrock_client(&mut self) -> Result<&BedrockClient, AppError> {
        if self.bedrock_client.is_none() {
            let aws_config = self.build_aws_config().await?;
            self.bedrock_client = Some(BedrockClient::new(&aws_config));
        }
        Ok(self.bedrock_client.as_ref().unwrap())
    }

    async fn build_aws_config(&self) -> Result<aws_config::SdkConfig, AppError> {
        let mut config_builder = aws_config::defaults(BehaviorVersion::v2025_01_17());

        if let Some(region) = &self.config.aws_bedrock_region {
            config_builder = config_builder
                .region(Region::new(region.clone()))
                .profile_name(&self.config.aws_bedrock_profile_name);
        }

        if let (Some(access_key), Some(secret_key)) = (
            &self.config.aws_bedrock_access_key_id,
            &self.config.aws_bedrock_secret_access_key,
        ) {
            let credentials = aws_sdk_bedrockruntime::config::Credentials::new(
                access_key,
                secret_key,
                None,
                None,
                "kate-chat",
            );
            config_builder = config_builder.credentials_provider(credentials);
        }

        Ok(config_builder.load().await)
    }

    fn get_model_provider(&self, model_id: &str) -> String {
        if model_id.starts_with("us.") || model_id.starts_with("eu.") || model_id.starts_with("ap.")
        {
            model_id.split('.').nth(1).unwrap_or("unknown").to_string()
        } else {
            model_id.split('.').next().unwrap_or("unknown").to_string()
        }
    }
}

#[async_trait]
impl AIProviderService for BedrockService {
    async fn invoke_model(&self, request: InvokeModelRequest) -> Result<ModelResponse, AppError> {
        let mut service = self.clone();
        let client = service.get_runtime_client().await?;

        let provider = self.get_model_provider(&request.model_id);

        let body = match provider.as_str() {
            "anthropic" => AnthropicProvider::format_request(&request)?,
            "amazon" => AmazonProvider::format_request(&request)?,
            "ai21" => AI21Provider::format_request(&request)?,
            "cohere" => CohereProvider::format_request(&request)?,
            "meta" => MetaProvider::format_request(&request)?,
            "mistral" => MistralProvider::format_request(&request)?,
            _ => {
                return Err(AppError::Validation(format!(
                    "Unsupported provider: {}",
                    provider
                )))
            }
        };

        let body_bytes = serde_json::to_vec(&body).map_err(|e| {
            error!(
                "Failed to serialize Bedrock request for model {}: {:?}",
                request.model_id, e
            );
            AppError::Internal(format!("Failed to serialize request: {}", e))
        })?;

        let response = client
            .invoke_model()
            .model_id(&request.model_id)
            .body(Blob::new(body_bytes))
            .send()
            .await
            .map_err(|e| {
                error!(
                    "Bedrock invoke failed for model {}: {:?}",
                    request.model_id, e
                );
                AppError::Aws(format!(
                    "Bedrock invoke failed: {}",
                    e.source().unwrap_or(&e).to_string()
                ))
            })?;

        let response_body = response.body().as_ref();
        let response_json: Value = serde_json::from_slice(response_body).map_err(|e| {
            error!(
                "Failed to parse Bedrock response for model {}: {:?}. Response body: {:?}",
                request.model_id,
                e,
                String::from_utf8_lossy(response_body)
            );
            AppError::Internal(format!("Failed to parse response: {}", e))
        })?;

        match provider.as_str() {
            "anthropic" => {
                AnthropicProvider::parse_model_response(response_json, &request.model_id)
            }
            "amazon" => AmazonProvider::parse_model_response(response_json, &request.model_id),
            "ai21" => AI21Provider::parse_model_response(response_json, &request.model_id),
            "cohere" => CohereProvider::parse_model_response(response_json, &request.model_id),
            "meta" => MetaProvider::parse_model_response(response_json, &request.model_id),
            "mistral" => MistralProvider::parse_model_response(response_json, &request.model_id),
            _ => Err(AppError::Validation(format!(
                "Unsupported provider: {}",
                provider
            ))),
        }
    }

    async fn invoke_model_stream<F, C, E>(
        &self,
        request: InvokeModelRequest,
        callbacks: StreamCallbacks<F, C, E>,
    ) -> Result<(), AppError>
    where
        F: Fn(String) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync,
        C: Fn(String) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync,
        E: Fn(AppError) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync,
    {
        // TODO: implement real streaming
        match self.invoke_model(request).await {
            Ok(response) => {
                // Simulate streaming by sending chunks
                let words: Vec<&str> = response.content.split_whitespace().collect();
                for word in words {
                    (callbacks.on_token)(format!("{} ", word)).await;
                    tokio::time::sleep(tokio::time::Duration::from_millis(10)).await;
                }
                (callbacks.on_complete)(response.content).await;
                Ok(())
            }
            Err(e) => {
                (callbacks.on_error)(e.clone()).await;
                Err(e)
            }
        }
    }

    async fn get_models(&self) -> Result<HashMap<String, AIModelInfo>, AppError> {
        let mut service = self.clone();
        let client = service.get_bedrock_client().await?;

        let response = client.list_foundation_models().send().await.map_err(|e| {
            error!("Failed to list Bedrock models: {:?}", e);
            AppError::Aws(format!("Failed to list models: {}", e))
        })?;

        let mut models = HashMap::new();

        for model in response.model_summaries() {
            if let (model_id, Some(model_name), Some(provider_name)) =
                (model.model_id(), model.model_name(), model.provider_name())
            {
                let supports_streaming = model.response_streaming_supported().unwrap_or(false);

                let supports_text_in = model
                    .input_modalities()
                    .iter()
                    .any(|mod_type| *mod_type == ModelModality::Text);
                let supports_image_in = model
                    .input_modalities()
                    .iter()
                    .any(|mod_type| *mod_type == ModelModality::Image);
                let supports_text_out = model
                    .output_modalities()
                    .iter()
                    .any(|mod_type| *mod_type == ModelModality::Text);
                let supports_image_out = model
                    .output_modalities()
                    .iter()
                    .any(|mod_type| *mod_type == ModelModality::Image);

                models.insert(
                    model_id.to_string(),
                    AIModelInfo {
                        api_provider: ApiProvider::AwsBedrock,
                        provider: Some(provider_name.to_string()),
                        name: model_name.to_string(),
                        description: format!("{} by {}", model_name, provider_name),
                        supports_streaming,
                        supports_text_in,
                        supports_text_out,
                        supports_image_in,
                        supports_image_out,
                        supports_embeddings_in: false,
                        supports_embeddings_out: false,
                    },
                );
            }
        }

        Ok(models)
    }

    async fn get_info(&self, test_connection: bool) -> Result<ProviderInfo, AppError> {
        let mut details = HashMap::new();
        details.insert("configured".to_string(), "true".to_string());

        if let Some(region) = &self.config.aws_bedrock_region {
            details.insert("region".to_string(), region.clone());
        }

        let is_connected =
            self.config.aws_bedrock_access_key_id.is_some() || std::env::var("AWS_PROFILE").is_ok();

        if test_connection && is_connected {
            let mut service = self.clone();
            match service.get_bedrock_client().await {
                Ok(_) => {
                    details.insert("connection_test".to_string(), "success".to_string());
                }
                Err(e) => {
                    details.insert("connection_test".to_string(), "failed".to_string());
                    details.insert("error".to_string(), e.to_string());
                }
            }
        }

        Ok(ProviderInfo {
            id: "aws_bedrock".to_string(),
            name: "AWS Bedrock".to_string(),
            is_connected,
            costs_info_available: is_connected,
            details,
        })
    }

    async fn get_costs(
        &self,
        start_time: i64,
        end_time: Option<i64>,
    ) -> Result<UsageCostInfo, AppError> {
        // TODO: Implement AWS Cost Explorer integration
        Ok(UsageCostInfo {
            start: DateTime::from_timestamp(start_time, 0).unwrap_or_default(),
            end: end_time.and_then(|t| DateTime::from_timestamp(t, 0)),
            costs: vec![],
            error: Some("Cost information not yet implemented".to_string()),
        })
    }
}

impl Clone for BedrockService {
    fn clone(&self) -> Self {
        Self {
            config: self.config.clone(),
            runtime_client: None,
            bedrock_client: None,
        }
    }
}
